core/src/main/scala/cats/instances/list.scala

```diff
@@ -120,50 +121,58 @@ trait ListInstances extends cats.kernel.instances.ListInstances {
       def traverse[G[_], A, B](fa: List[A])(f: A => G[B])(implicit G: Applicative[G]): G[List[B]] =
         if (fa.isEmpty) G.pure(Nil)
         else
-          G.map(Chain.traverseViaChain {
-            val as = collection.mutable.ArrayBuffer[A]()
-            as ++= fa
-            wrapMutableIndexedSeq(as)
-          }(f))(_.toList)
+          G match {
+            case x: StackSafeMonad[G] => x.map(Traverse.traverseDirectly[G, A, B](fa)(f)(x))(_.toList)
+            case _ =>
+              G.map(Chain.traverseViaChain {
+                val as = collection.mutable.ArrayBuffer[A]()
+                as ++= fa
+                wrapMutableIndexedSeq(as)
+              }(f))(_.toList)
+          }
 
       /**
        * This avoids making a very deep stack by building a tree instead
        */
       override def traverse_[G[_], A, B](fa: List[A])(f: A => G[B])(implicit G: Applicative[G]): G[Unit] = {
-        // the cost of this is O(size log size)
-        // c(n) = n + 2 * c(n/2) = n + 2(n/2 log (n/2)) = n + n (logn - 1) = n log n
-        // invariant: size >= 1
-        def runHalf(size: Int, fa: List[A]): Eval[G[Unit]] =
-          if (size > 1) {
-            val leftSize = size / 2
-            val rightSize = size - leftSize
-            val (leftL, rightL) = fa.splitAt(leftSize)
-            runHalf(leftSize, leftL)
-              .flatMap { left =>
-                val right = runHalf(rightSize, rightL)
-                G.map2Eval(left, right) { (_, _) => () }
+        G match {
+          case x: StackSafeMonad[G] => Traverse.traverse_Directly(fa)(f)(x)
+          case _                    =>
+            // the cost of this is O(size log size)
+            // c(n) = n + 2 * c(n/2) = n + 2(n/2 log (n/2)) = n + n (logn - 1) = n log n
+            // invariant: size >= 1
+            def runHalf(size: Int, fa: List[A]): Eval[G[Unit]] =
+              if (size > 1) {
+                val leftSize = size / 2
+                val rightSize = size - leftSize
+                val (leftL, rightL) = fa.splitAt(leftSize)
+                runHalf(leftSize, leftL)
+                  .flatMap { left =>
+                    val right = runHalf(rightSize, rightL)
```

Comment:
an interesting benchmark would be do do `Eval.defer(runHalf(rightSize, rightL))` which would short circuit for applicatives G that can short circuit.

I wonder how much that would hurt the non-short circuiting case... this is not for this PR, just something we may come back to in a later PR.